chore: package addons `Fn::Transform`

dannyrandall opened the pull request:
This adds support for packaging the `Location` parameter on a [`Fn::Transform`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-transform.html) that uses the [`AWS::Include`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/create-reusable-transform-function-snippets-and-add-to-your-template-with-aws-include-transform.html) macro.

This also fixes a small bug in `s3.ParseURL()` where we don't correctly parse bucket names with `.`'s in it.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

dannyrandall committed:
moved over `Fn::Transform` stuff

dannyrandall committed:
ok saving this version even though it doesn't save comments

dannyrandall committed:
added back version that works with comments

dannyrandall committed:
support s3 bucket names with dots

dannyrandall committed:
support multiple nodes in `packageTransforms()`
test for mapping/sequence nodes

dannyrandall committed:
remove TODO

dannyrandall committed:
package error test cases

dannyrandall committed:
Merge branch 'mainline' into chore/addons-fn-transform

efekarakus commented on internal/pkg/addon/package.go:
```diff
@@ -193,9 +191,13 @@ var resourcePackageConfig = map[string][]packagePropertyConfig{
 
 func (t *cfnTemplate) pkg(a *Addons) error {
-	resources := mappingNode(&t.Resources)
+	err := a.packageTransforms(t.Metadata, t.Parameters, t.Mappings, t.Conditions, t.Transform, t.Resources, t.Outputs)
```
> We can use the AWS::Include in any section of the template, except Parameters. 
https://medium.com/swlh/using-the-cloudformation-aws-include-macro-9e3056cf75b0

This is such a funky transformation, I wonder how many people use it

dannyrandall replied:
what's extra weird to me is how the `aws cloudformation package` doesn't package an `AWS::Include` if it's the top level transform 🤔 
like:
```
Transform:
  Name: AWS::Include
  Parameters:
    Location: ./tags
Resources:
# blah
Outputs:
# bleh
```
just leaves `Transform.Parameters.Location` as is :joy:

efekarakus commented on internal/pkg/addon/package.go:
```diff
@@ -213,6 +215,60 @@ func (t *cfnTemplate) pkg(a *Addons) error {
+// dectects one, and the "Location" parameter is set to a local path, it'll
+// upload those files to S3. If node is a yaml map or sequence, it will
+// recursivly traverse those nodes.
+func (a *Addons) packageTransforms(nodes ...yaml.Node) error {
```
nit: What do you think of renaming it to?
```suggestion
func (a *Addons) packageIncludeTransforms(roots ...yaml.Node) error {
```

efekarakus commented on internal/pkg/addon/package.go:
```diff
@@ -213,6 +215,60 @@ func (t *cfnTemplate) pkg(a *Addons) error {
 
+// packageTransforms searches each node in nodes for the CFN
+// intrinsic function "Fn::Transform" with the "AWS::Include" macro. If it
+// dectects one, and the "Location" parameter is set to a local path, it'll
```
```suggestion
// detects one, and the "Location" parameter is set to a local path, it'll
```

efekarakus commented on internal/pkg/addon/package.go:
```diff
@@ -213,6 +215,60 @@ func (t *cfnTemplate) pkg(a *Addons) error {
+// intrinsic function "Fn::Transform" with the "AWS::Include" macro. If it
+// dectects one, and the "Location" parameter is set to a local path, it'll
+// upload those files to S3. If node is a yaml map or sequence, it will
+// recursivly traverse those nodes.
```
```suggestion
// recursively traverse those nodes.
```
doing @kohidave proud with the typos

dannyrandall replied:
😇 i'm just impressed that you catch them

efekarakus commented on internal/pkg/addon/package.go:
```diff
@@ -213,6 +215,60 @@ func (t *cfnTemplate) pkg(a *Addons) error {
+// upload those files to S3. If node is a yaml map or sequence, it will
+// recursivly traverse those nodes.
+func (a *Addons) packageTransforms(nodes ...yaml.Node) error {
+	forNode := func(node yaml.Node) error {
```
nit: Should we take as input a pointer to make it clear that we're mutating the internals
```suggestion
	pkg := func(node *yaml.Node) error {
```

efekarakus commented on internal/pkg/aws/s3/s3.go:
```diff
@@ -153,8 +153,20 @@ func ParseURL(url string) (bucket string, key string, err error) {
+		}
+	}
+
+	return strings.Join(split[:bucketEndIdx], "."), parsedURL[1], nil
```
this definitely threw me off 😂 but great test case to make it clear what's happening

paragbhingre replied:
I like this super amazing logic ❤️
But how about this? 
```suggestion
	bucket, key = strings.Split(parsedURL[0], ".s3")[0], parsedURL[1]
```

It's basically the same logic as before but just adding `.s3` instead of `.` gives us the result that we want. 
I don't why I can't suggest code block for the removed code up there?

paragbhingre replied:
nvm if customers put `.s3` in their name itself then logic fails. So what you have is good 💯

efekarakus reviewed (approved):
Looks great, just small things

dannyrandall committed:
address fb:
- don't package include transforms in `Parameters`
- spell better 😅
- rename functions
- accept `*yaml.Node` instead of `yaml.Node`

dannyrandall committed:
smh

dannyrandall committed:
add test for s3 current version of URL

paragbhingre reviewed (approved):
Ship it 🎉

mergify[bot] committed:
Merge branch 'mainline' into chore/addons-fn-transform
